### PR TITLE
A rough cut of a working Bulk API executor and return values

### PR DESCRIPTION
**Pull request by jon-betts**
### What's in this PR?

 * This adds implementation guts for actually writing all of the required changes to the database for the Bulk API
    * A call to this end-point should really upsert and create everything
    * It's very happy path though
 * This is done with 'action' classes that each concentrate on doing one thing (like `UserUpsertAction`). This is mostly to prevent the executor from becoming crazy long. These are in `h.services.bulk_executor.actions`.
 * A return data path has been implemented with a new view type "basic". This will return the data type and the public id of an object. This is for debugging and comfort only. We'll use view type `None` when we are doing this for real
 * All fakery has been stripped out of the classes

### What's not in this PR?

 * Some totally separable things have already been split out into: https://github.com/hypothesis/h/pull/5987
 * Incomplete doc strings
 * Probably quite a bit of missing error handing and edge-cases (do we actually abort the transaction if something fails etc?)
 * Lots of un-implemented tests (and general coverage gaps)

----

### Reviewing notes

This PR is intended as a way to check in on the general direction, and therefore:

 * This is far too big to actually review properly (it is intended to be chopped up soon)
 * The diff might not be worth reading in detail
 * The files as they are laid out are probably more useful

Things which are mostly noise here:

 * The service has been split into multiple files to make it easier to read
 * Lots of places changing from `example.com` to `lms.hypothes.is` for the authority, as it's the only one we accept

A working call with dev data:

```
curl --request POST \
  --url http://localhost:5000/api/bulk \
  --header 'authorization: Basic ZjVhZGE5NzItNzMyZC0xMWU5LTk2NzUtNzMyNjhlYjlkNzYxOjQ1dHZlZF8ya3NnZlNMYjFEUlYzWktNbHo1c2tWVFRQaXJyQ3lKaUlUam8=' \
  --data '["configure", {"view": "basic", "user": {"effective": "acct:3a022b6c146dfd9df4ea8662178eac@lms.hypothes.is"}, "instructions": {"total": 4}, "defaults": [["create", "*", {"on_duplicate": "continue"}], ["upsert", "*", {"merge_query": true}]]}]
["upsert", {"data": {"type": "user", "attributes": {"display_name": "display name2", "identities": [{"provider": "provider string", "provider_unique_id": "provider unique id2"}]}, "meta": {"query": {"authority": "lms.hypothes.is", "username": "username2"}, "$anchor": "user_ref"}}}]
["upsert", {"data": {"type": "group", "attributes": {"name": "name"}, "meta": {"query": {"authority": "lms.hypothes.is", "authority_provided_id": "authority_provided_id"}, "$anchor": "group_ref"}}}]
["create", {"data": {"type": "group_membership", "relationships": {"member": {"data": {"type": "user", "id": {"$ref": "user_ref"}}}, "group": {"data": {"type": "group", "id": {"$ref": "group_ref"}}}}}}]
```

### CODE REVIEW

File: h/services/bulk_executor/executors.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+        if handler is None:
+            raise UnsupportedOperationError(
+                f"No implementation for {command_type.value} {data_type.value}"
+            )
+
+        return handler
+
+
+class AuthorityCheckingExecutor(DBExecutor):
```
Comment (seanh) [inline h/services/bulk_executor/executors.py]: I think it'd be nicer to use compostion over inheritance.

This class isn't really an authority-checking executor. It's an authority-checking-and-also-DB-writing-executor.

One possible approach might be:

```python
class ComboExecutor:
    def __init__(self, executors):
        self._executors = executors

    def configure(self, *args, **kwargs):
        for executor in self._executors:
            executor.configure(*args, **kwargs)

    def execute_batch(self, *args, **kwargs):
        for executor in self._executors:
            executor.execute_batch(*args, **kwargs)

def bulk_executor_service_factory(_, request):
    return ComboExecutor([AuthorityCheckingExecutor(), DBExecutor(request.db)])
```

`AuthorityCheckingExecutor` then wouldn't need the DB, inheritance or super calls.

Comment (seanh) [reply h/services/bulk_executor/executors.py]: `AuthorityCheckingExecutor` code should be simpler to understand as you no longer need to think about `DBExecutor` to understand it. With inheritance `AuthorityCheckingExecutor` has a `db` param that's actually `DBExecutor`'s param, for example, and it knows about `DBExecutor.__init__()`'s signature.

This would also mean that `AuthorityCheckingExecutor` could be unit-tested without also executing `DBExecutor`'s code and touching the DB (with inheritance, you can't mock out the base class so you can't test `AuthorityCheckingExecutor` without `DBExecutor`).

Similarly `ComboExecutor` can be both understood and unit-tested without having to know anything about the other two executors.

It's more flexible too, although we don't need that flexibility yet. Currently the only sequence possible is `AuthorityCheckingExecutor` -> `DBExecutor`, as that's hardcoded in the inheritance. But `ComboExecutor` can combine any set of executors in any order.

File: h/services/bulk_executor/actions.py [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+
+        # Check that we can actually process this batch
+        self._check_upsert_queries(
+            batch, expected_keys=["authority", "authority_provided_id"]
+        )
+
+        # Prep the query
+        values = [command.body.attributes for command in batch]
+        for value in values:
+            value["creator_id"] = effective_user_id
```
[the remark this answers is not in the document]
Comment (jon-betts) [reply h/services/bulk_executor/actions.py]: Yeah, I spotted this while looking through other code. To support the existing behavior I think we would need a separate `GroupUpdate` with `on_duplicate=fail` for the student. If we can get away without doing that by waiting though... that would be good.

It's also a cheeky reason to delay merging into LMS etc.

File: h/services/bulk_executor/actions.py [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+
+        return Report(Report.CommandResult.UPSERTED, id_, public_id=groupid)
+
+
+class GroupMembershipCreateAction(DBAction):
+    def execute(self, batch, on_duplicate="continue", **_):
+        if on_duplicate != "continue":
+            raise UnsupportedOperationError(
+                "Create modes other than 'continue' have not been implemented"
+            )
```
[the remark this answers is not in the document]
Comment (jon-betts) [reply h/services/bulk_executor/actions.py]: As noted above, it's a likely future requirement. This means if we did implement a new mode and called an old deployment it would fail sensibly.

File: h/services/bulk_executor/service.py [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+
+class BulkExecutorService:
+    """Constructs executors for the Bulk API view."""
+
+    def __init__(self, context, request):
+        self.db = request.db
+
+    def get_executor(self):
+        """Get an Executor for carrying out bulk actions."""
+        return AuthorityCheckingExecutor(self.db)
```
[the remark this answers is not in the document]
Comment (jon-betts) [reply h/services/bulk_executor/service.py]: I'm a bit confused about this, as _everything_ has a service factory in H, but it's just a function. 

I agree it's circuitous, but I think that about the service pattern in general. I think returning an executor as if it's a singleton, when they definitely aren't isn't a good idea. I'm 100% up for doing (2) here.

File: h/services/bulk_executor/executors.py (hunk 1 above)
Comment (jon-betts) [reply h/services/bulk_executor/executors.py]: The other thing I think I might do, is just merge the two together. The whole hot-swapping executors thing strikes me as YAGNI. I don't think we'll ever have another behavior.

File: h/services/bulk_executor/actions.py [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+from h.models.group import Group, GroupMembership
+from h.models.user import User
+
+
+class DBAction:
+    def __init__(self, db):
+        self.db = db
+
+    def execute(self, batch, **options):
+        raise NotImplementedError()
```
[the remark this answers is not in the document]
Comment (jon-betts) [reply h/services/bulk_executor/actions.py]: I'm not sure exactly what's being proposed here. 

While there is only one effective _public_ method on these objects, there are plenty of private methods to break up the problem. We could explode them all out into a bunch of free floating methods in the package. But then it wouldn't be clear which functions belong to what.

For those private methods, you can pass the db argument as the first argument to each of them, but if you are repeatedly passing the same argument to every function in a collection of functions which only work together... you should probably use a class?

File: h/views/api/bulk.py [hunk 6]
```diff
@@ -90,9 +33,9 @@ def bulk(request):
     `h.h_api.bulk_api`.
     """
 
-    results = FakeBulkAPI.from_byte_stream(
-        request.body_file, executor=AuthorityCheckingExecutor()
-    )
+    executor = request.find_service(name="bulk_executor").get_executor()
+
+    results = BulkAPI.from_byte_stream(request.body_file, executor=executor)
```
[the remark this answers is not in the document]
Comment (jon-betts) [reply h/views/api/bulk.py]: There might be some scope to suck up some of the later stuff in there too. I'll take a look.

File: h/services/bulk_executor/executors.py (hunk 1 above)
Comment (jon-betts) [reply h/services/bulk_executor/executors.py]: It also just occurred to me (facepalm) that if there's only one authority, we can enforce it in the schema and avoid all of this?

It's not as "flexible" but why worry about that? This would have the benefit of being more specific about what you got wrong too.

Comment (jon-betts) [reply h/services/bulk_executor/executors.py]: Gah... me from the past/future. The reason this was decided against is it's not the libraries job to police authentication style issues, that's should reside in H.

File: h/views/api/bulk.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (jon-betts) [reply h/views/api/bulk.py:38]: I tried combining this stuff together but it leads to a conceptual problem. Currently the majority of valdiation and enforcement of the structure is performed by BulkAPI. The actual database manipulation is done in the executor (plus some necessary checking there too).

This means most of the 'view' like behavior is in BulkAPI, and most of the 'service' like behavior is in the executor. 

I think this view is small enough to fit on a screen and wouldn't benefit by chopping up any more, so I'll leave this split in place.

What I have done is remove any pyramid service trappings from it, it's just fine to import and use it. We don't want a singleton, we use it once and chuck it away.

File: h/services/bulk_executor/executors.py (hunk 1 above)
Comment (jon-betts) [reply h/services/bulk_executor/executors.py:57]: Ok, so I've come down on the side of merging the two which means there is no more inheritance problem to solve. The classes' job is to prep for execution and do it.

It's smaller than the previous two and probably doesn't merit chopping up much.

**Comment by jon-betts** [comment]
Ok so with regards to the three major issues raised:

 * The executor is no longer a pyramid service or provided by one, you just import it
 * The two executors are now one, avoiding any inheritance problems
 * I tried the actions classes a few different ways, but the user action is 80 lines long right now, which I think is too much for a single function. I'm going to leave these as they are.

**Comment by jon-betts** [comment]
Nothing left here it's all been parcelled out
